File: common/src/main/java/org/sandboxpowered/silica/network/play/clientbound/PlayerInfo.java
package org.sandboxpowered.silica.network.play.clientbound;

import com.mojang.authlib.GameProfile;
import com.mojang.authlib.properties.PropertyMap;
import org.sandboxpowered.silica.network.PacketByteBuf;
import org.sandboxpowered.silica.network.PacketHandler;
import org.sandboxpowered.silica.network.PacketPlay;
import org.sandboxpowered.silica.network.PlayContext;

import java.util.UUID;

public class PlayerInfo implements PacketPlay {
    private int action;

    private UUID[] uuids;

    private String[] names; // Action 0
    private PropertyMap[] propertyMaps; // Action 0
    private int[] gamemodes; // Action 0,1
    private int[] pings; // Action 0,2

    public PlayerInfo() {
    }

    public PlayerInfo(int action, UUID[] uuids, String[] names, PropertyMap[] propertyMaps, int[] gamemodes, int[] pings) {
        this.action = action;
        this.uuids = uuids;
        this.names = names;
        this.propertyMaps = propertyMaps;
        this.gamemodes = gamemodes;
        this.pings = pings;
    }

    public static PlayerInfo addPlayer(GameProfile[] profiles, int[] gamemodes, int[] pings) {
        UUID[] uuids = new UUID[profiles.length];
        String[] names = new String[profiles.length];
        PropertyMap[] propertyMaps = new PropertyMap[profiles.length];
        for (int i = 0; i < profiles.length; i++) {
            GameProfile profile = profiles[i];
            uuids[i] = profile.getId();
            names[i] = profile.getName();
            propertyMaps[i] = profile.getProperties();
        }
        return new PlayerInfo(0, uuids, names, propertyMaps, gamemodes, pings);
    }

    public static PlayerInfo updateLatency(UUID[] uuids, int[] pings) {
        return new PlayerInfo(2, uuids, new String[0], new PropertyMap[0], new int[0], pings);
    }

    public static PlayerInfo removePlayer(UUID[] uuids) {
        return new PlayerInfo(4, uuids, new String[0], new PropertyMap[0], new int[0], new int[0]);
    }

    @Override
    public void read(PacketByteBuf buf) {
        action = buf.readVarInt();
    }

    @Override
    public void write(PacketByteBuf buf) {
        buf.writeVarInt(action);
        buf.writeVarInt(uuids.length);

        for (int i = 0; i < uuids.length; i++) {
            buf.writeUUID(uuids[i]);
            if (action == 0) {
                buf.writeString(names[i], 16);
                PropertyMap map = propertyMaps[i];
                buf.writeVarInt(map.size());
                map.forEach((s, property) -> {
                    buf.writeString(property.getName());
                    buf.writeString(property.getValue());
                    buf.writeBoolean(property.hasSignature());
                    if (property.hasSignature()) {
                        buf.writeString(property.getSignature());
                    }
                });
                buf.writeVarInt(gamemodes[i]);
                buf.writeVarInt(pings[i]);
                buf.writeBoolean(false);
            } else if (action == 1) {
                buf.writeVarInt(gamemodes[i]);
            } else if (action == 2) {
                buf.writeVarInt(pings[i]);
            } else if (action == 3) {
                buf.writeBoolean(false);
            }
        }
    }

    @Override
    public void handle(PacketHandler packetHandler, PlayContext context) {

    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/network/play/clientbound/PlayerInfo.kt
package org.sandboxpowered.silica.network.play.clientbound

import com.mojang.authlib.GameProfile
import com.mojang.authlib.properties.Property
import com.mojang.authlib.properties.PropertyMap
import org.sandboxpowered.silica.network.PacketByteBuf
import org.sandboxpowered.silica.network.PacketHandler
import org.sandboxpowered.silica.network.PacketPlay
import org.sandboxpowered.silica.network.PlayContext
import java.util.*

class PlayerInfo(
    private var action: Int = -1,
    private var uuids: Array<UUID> = emptyArray(),
    private var names: Array<String> = emptyArray(),// Action 0
    private var propertyMaps: Array<PropertyMap> = emptyArray(),// Action 0
    private var gamemodes: IntArray = intArrayOf(),// Action 0,1
    private var pings: IntArray = intArrayOf(),// Action 0,2
) : PacketPlay {
    override fun read(buf: PacketByteBuf) {
        action = buf.readVarInt()
    }

    override fun write(buf: PacketByteBuf) {
        buf.writeVarInt(action)
        buf.writeVarInt(uuids.size)
        for (i in uuids.indices) {
            buf.writeUUID(uuids[i])
            when (action) {
                0 -> {
                    buf.writeString(names[i], 16)
                    val map = propertyMaps[i]
                    buf.writeVarInt(map.size())
                    map.forEach { _, property: Property ->
                        buf.writeString(property.name)
                        buf.writeString(property.value)
                        buf.writeBoolean(property.hasSignature())
                        if (property.hasSignature()) {
                            buf.writeString(property.signature)
                        }
                    }
                    buf.writeVarInt(gamemodes[i])
                    buf.writeVarInt(pings[i])
                    buf.writeBoolean(false)
                }
                1 -> buf.writeVarInt(gamemodes[i])
                2 -> buf.writeVarInt(pings[i])
                3 -> buf.writeBoolean(false)
            }
        }
    }

    override fun handle(packetHandler: PacketHandler, context: PlayContext) {}

    companion object {
        fun addPlayer(profiles: Array<GameProfile>, gamemodes: IntArray, pings: IntArray): PlayerInfo {
            val uuids = arrayOfNulls<UUID>(profiles.size)
            val names = arrayOfNulls<String>(profiles.size)
            val propertyMaps = arrayOfNulls<PropertyMap>(profiles.size)
            for (i in profiles.indices) {
                val profile = profiles[i]
                uuids[i] = profile.id
                names[i] = profile.name
                propertyMaps[i] = profile.properties
            }
            return PlayerInfo(
                0,
                uuids.requireNoNulls(),
                names.requireNoNulls(),
                propertyMaps.requireNoNulls(),
                gamemodes,
                pings
            )
        }

        fun updateLatency(uuids: Array<UUID?>, pings: IntArray): PlayerInfo {
            return PlayerInfo(2, uuids.requireNoNulls(), emptyArray(), emptyArray(), IntArray(0), pings)
        }

        fun removePlayer(uuids: Array<UUID?>): PlayerInfo {
            return PlayerInfo(4, uuids.requireNoNulls(), emptyArray(), emptyArray(), intArrayOf(), intArrayOf())
        }
    }
}